File: mobile/lib/features/fuel/domain/fuel_entry_calc.dart
enum FuelField { liters, total, perLiter }

/// Resolves the liters / total-amount / price-per-liter triple for quick fuel
/// entry. Only two of the three are needed; the third is derived. Price-per-
/// liter starts as a sticky default (prefilled from the last log).
///
/// The two most-recently-edited fields are authoritative; the remaining field
/// is computed. When per-liter is edited while both others are set, liters is
/// kept and total is recomputed.
///
/// [derivedField] exposes which of the three fields is currently being computed
/// so the UI can tag it as "AUTO". When fewer than two fields are set the
/// derived field is null (no computation possible yet).
class FuelEntryCalc {
  double? liters;
  double? total;
  double? pricePerLiter;

  final List<FuelField> _locked = [];

  FuelEntryCalc({double? initialPerLiter}) {
    if (initialPerLiter != null && initialPerLiter > 0) {
      pricePerLiter = initialPerLiter;
      _locked.add(FuelField.perLiter);
    }
  }

  /// The field currently being derived (AUTO), or null when there isn't one.
  FuelField? get derivedField {
    if (_locked.length < 2) return null;
    return FuelField.values.firstWhere((f) => !_locked.contains(f));
  }

  void setField(FuelField field, double? value) {
    switch (field) {
      case FuelField.liters:
        liters = value;
      case FuelField.total:
        total = value;
      case FuelField.perLiter:
        pricePerLiter = value;
    }

    _locked.remove(field);
    if (value == null) {
      _recompute();
      return;
    }
    _locked.add(field);
    while (_locked.length > 2) {
      final victim =
          (field == FuelField.perLiter && _locked.contains(FuelField.total))
          ? FuelField.total
          : _locked.first;
      _locked.remove(victim);
    }
    _recompute();
  }

  void _recompute() {
    if (_locked.length < 2) return;
    final computed = FuelField.values.firstWhere((f) => !_locked.contains(f));
    switch (computed) {
      case FuelField.perLiter:
        if (liters != null && liters! > 0 && total != null) {
          pricePerLiter = total! / liters!;
        }
      case FuelField.liters:
        if (total != null && pricePerLiter != null && pricePerLiter! > 0) {
          liters = total! / pricePerLiter!;
        }
      case FuelField.total:
        if (liters != null && pricePerLiter != null && pricePerLiter! > 0) {
          total = liters! * pricePerLiter!;
        }
    }
  }

  bool get isComplete =>
      liters != null && liters! > 0 && total != null && total! >= 0;
}

File: mobile-rn/src/features/fuel-logs/fuel-entry-calc.ts
/**
 * Resolves the liters / total / price-per-liter triple for quick fuel entry.
 * Parity with Flutter `features/fuel/domain/fuel_entry_calc.dart`.
 *
 * Only two of the three are needed; the third is derived. The two most-recently
 * edited fields are authoritative. Price-per-liter starts as a sticky default
 * (prefilled from the last log).
 *
 * `derivedField` exposes which field is currently computed so the UI can tag it
 * "AUTO". With fewer than two fields set there is no derived field.
 */

export type FuelField = 'liters' | 'total' | 'perLiter';

/** Declaration order matters: Flutter iterates `FuelField.values` to find the
 *  single unlocked field, so this array must stay liters → total → perLiter. */
const FUEL_FIELDS: FuelField[] = ['liters', 'total', 'perLiter'];

export class FuelEntryCalc {
  liters: number | null = null;
  total: number | null = null;
  pricePerLiter: number | null = null;

  /** Ordered list of the most-recently-edited fields; at most two. */
  private locked: FuelField[] = [];

  constructor(initialPerLiter?: number | null) {
    if (initialPerLiter != null && initialPerLiter > 0) {
      this.pricePerLiter = initialPerLiter;
      this.locked.push('perLiter');
    }
  }

  /** The field currently being derived (AUTO), or null when there isn't one. */
  get derivedField(): FuelField | null {
    if (this.locked.length < 2) return null;
    return FUEL_FIELDS.find((f) => !this.locked.includes(f)) ?? null;
  }

  setField(field: FuelField, value: number | null): void {
    switch (field) {
      case 'liters':
        this.liters = value;
        break;
      case 'total':
        this.total = value;
        break;
      case 'perLiter':
        this.pricePerLiter = value;
        break;
    }

    this.removeLocked(field);

    // Clearing a field unlocks it without making it authoritative.
    if (value === null) {
      this.recompute();
      return;
    }

    this.locked.push(field);

    while (this.locked.length > 2) {
      // Editing per-liter with total set evicts total (liters is kept and total
      // recomputed) — otherwise evict the oldest locked field.
      const victim =
        field === 'perLiter' && this.locked.includes('total') ? 'total' : this.locked[0];
      this.removeLocked(victim);
    }

    this.recompute();
  }

  private removeLocked(field: FuelField): void {
    const index = this.locked.indexOf(field);
    if (index !== -1) this.locked.splice(index, 1);
  }

  private recompute(): void {
    if (this.locked.length < 2) return;

    const computed = FUEL_FIELDS.find((f) => !this.locked.includes(f));

    switch (computed) {
      case 'perLiter':
        if (this.liters != null && this.liters > 0 && this.total != null) {
          this.pricePerLiter = this.total / this.liters;
        }
        break;
      case 'liters':
        if (this.total != null && this.pricePerLiter != null && this.pricePerLiter > 0) {
          this.liters = this.total / this.pricePerLiter;
        }
        break;
      case 'total':
        if (this.liters != null && this.pricePerLiter != null && this.pricePerLiter > 0) {
          this.total = this.liters * this.pricePerLiter;
        }
        break;
    }
  }

  get isComplete(): boolean {
    return this.liters != null && this.liters > 0 && this.total != null && this.total >= 0;
  }
}
